pkg/square/square.go

```diff
@@ -42,45 +38,62 @@ func Construct(txs [][]byte, maxSquareSize int) (Square, [][]byte, error) {
 }
 
-// Reconstruct takes a list of ordered transactions and reconstructs a square, validating that
-// all PFBs are ordered after regular transactions and that the transactions don't collectively
-// exceed the maxSquareSize. Note that this function does not check the underlying validity of
+// Construct takes the exact list of ordered transactions and constructs a square, validating that
+//   - all PFBs are ordered after regular transactions that
+//   - the transactions don't collectively exceed the maxSquareSize.
+//
+// Note that this function does not check the underlying validity of
 // the transactions.
-func Reconstruct(txs [][]byte, maxSquareSize int) (Square, error) {
-	builder, err := NewBuilder(maxSquareSize)
+func Construct(txs [][]byte, maxSquareSize int) (Square, error) {
+	builder, err := NewBuilder(maxSquareSize, txs...)
 	if err != nil {
 		return nil, err
 	}
-	seenFirstBlobTx := false
-	for idx, tx := range txs {
-		blobTx, isBlobTx := core.UnmarshalBlobTx(tx)
-		if isBlobTx {
-			seenFirstBlobTx = true
-			canAppend, err := builder.AppendBlobTx(blobTx)
-			if err != nil {
-				return nil, err
-			}
-			if !canAppend {
-				return nil, fmt.Errorf("not enough space to append blob tx at index %d", idx)
-			}
-		} else {
-			if seenFirstBlobTx {
-				return nil, fmt.Errorf("normal tx at index %d can not be Appended after blob tx", idx)
-			}
-			if !builder.AppendTx(tx) {
-				return nil, fmt.Errorf("not enough space to append tx at index %d", idx)
-			}
-		}
-	}
 	return builder.Export()
 }
 
+// TxShareRange returns the range of share indexes that the tx, specified by txIndex, occupies.
+// Both ends of the range are inclusive.
+func TxShareRange(txs [][]byte, txIndex int) (shares.ShareRange, error) {
+	builder, err := NewBuilder(appconsts.DefaultMaxSquareSize, txs...)
+	if err != nil {
+		return shares.ShareRange{}, err
+	}
+
+	return builder.FindTxShareRange(txIndex)
+}
+
+// BlobShareRange returns the range of share indexes that the blob, identified by txIndex and blobIndex, occupies.
+// Both ends of the range are inclusive.
+func BlobShareRange(txs [][]byte, txIndex, blobIndex int) (shares.ShareRange, error) {
+	builder, err := NewBuilder(appconsts.DefaultMaxSquareSize, txs...)
+	if err != nil {
+		return shares.ShareRange{}, err
+	}
+
+	start, err := builder.FindBlobStartingIndex(txIndex, blobIndex)
+	if err != nil {
+		return shares.ShareRange{}, err
+	}
+
+	blobLen, err := builder.BlobShareLength(txIndex, blobIndex)
+	if err != nil {
+		return shares.ShareRange{}, err
+	}
+
+	return shares.ShareRange{Start: int(start), End: int(start) + blobLen - 1}, nil
+}
+
 // Square is a 2D square of shares with symmetrical sides that are always a power of 2.
 type Square []shares.Share
 
 // Size returns the size of the sides of a square
 func (s Square) Size() uint64 {
-	return uint64(math.Sqrt(float64(len(s))))
+	return Size(len(s))
+}
+
+func Size(len int) uint64 {
+	return uint64(math.Sqrt(float64(len)))
 }
```

Comment:
[not blocking] this implementation hasn't changed but I'm struggling to see why it was like this originally given https://go.dev/play/p/KQDnmVJHpni

I feel like it needs to round up

Reply:
I was assuming that the length was always a power of two but that might not be the case. I think you're right that it's safer if we round up